content/articles/statistical-hypothesis-testing-python-implementation/index.md

```diff
@@ -0,0 +1,520 @@
+ 
+![chi_square_formula](/engineering-education/statistical-hypothesis-testing-python-implementation/chi-square.PNG)
+ 
+Consider the following example:
+ 
+A mathematics teacher conducted a test for a class of 20 students. He expected 5 to score good marks, 7 to score average marks while 8 to fail. Eventually, when the result came out, 5 failed, 6 were average while 9 scored good marks. Mathematically, carry out chi-square test for this class.
+ 
+|  | Good  |  Average  | Fail | Total |
+|----|--------|:----------|--------------|--------|
+| Observed| 9 | 6 | 5 | 20 |
+| Expected| 5 | 7 | 8 | 20 |
+ 
+Where
+ 
+- observed good is A,
+- expected good is a,
+- observed average is B,
+- expected average is b,
+- observed fail is C, and
+- expected to fail is c.
+```
+Chi-square = Σ[(observed value - expected value)² / expected value]
+```
+```
+A-a = 4
+B-b = -1
+C-c = -3
+```
+```
+# find the square of our values
+/A-a/² = 16
+/B-b/² = 1
+/C-c/² = 9
+```
+```
+Good = (16/5) = 3.2
+Average = (1/7) = 0.143
+Fail = (9/8) = 1.125
+ 
+chi-square = 3.2 + 0.143 + 1.125
+chi-square = 4.468
+```
+For huge datasets, mathematical calculations cannot be done by hand. The implementation of the chi-square test in Python is discussed in the next section.
+ 
+```python
+import pandas as pd
+import numpy as np
+import scipy.stats as stats
+from scipy.stats import f_oneway
+from scipy.stats import chi2_contingency
+import matplotlib.pyplot as plt
+ 
+import statsmodels.api as sm
+from statsmodels.formula.api import ols
+ 
+import warnings
+warnings.filterwarnings( "ignore" )
+```
+ 
+The libraries imported are the most important libraries used for any data science python implementation work, except for the f_oneway used for the implementing one way ANOVA test and chi2_contigency used for the implementation of Chi-square.
+ 
+I will be using the popular iris data set for the tutorial, you can get it almost anywhere but I downloaded this from [Kaggle](https://www.kaggle.com/vikrishnan/iris-dataset).
+ 
+Load the data into the notebook
+```python
+df = pd.read_csv("/content/drive/MyDrive/IRIS.csv")
+```
+```python
+df.head()
+```
+![df.head.png](/engineering-education/statistical-hypothesis-testing-python-implementation/dataset_head.PNG)
+On checking the first five, it seems all the species of flower are setosa, but the iris dataset has 2 more samples apart from that, to be sure, let's run df.sample() to visualize samples randomly.
+```python
+df.sample(10)
+```
+![sampleten](/engineering-education/statistical-hypothesis-testing-python-implementation/dataset_sample.PNG)
+ 
+Always check for missing values to avoid future complications.
+```python
+df.isnull().sum()
+```
+```bash
+sepal_length    0
+sepal_width     0
+petal_length    0
+petal_width     0
+species         0
+dtype: int64
+```
+We can be sure now that there are 3 classes of species and there are no missing values.
```

Comment:
Do justify this statement by adding a piece of code (df["species"].unique()) before the statement to print the three classes of species.

Reply:
Please add a print statement so that the reader can see the output. You've just copied it. The right one would be -
```py
print(df["species"].unique())
```